oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/OAuth2AuthorizationCodeGrantFilter.java

```diff
@@ -145,16 +147,45 @@ private boolean shouldProcessAuthorizationResponse(HttpServletRequest request) {
 		if (authorizationRequest == null) {
 			return false;
 		}
-		String requestUrl = UrlUtils.buildFullRequestUrl(request.getScheme(), request.getServerName(),
-				request.getServerPort(), request.getRequestURI(), null);
 		MultiValueMap<String, String> params = OAuth2AuthorizationResponseUtils.toMultiMap(request.getParameterMap());
-		if (requestUrl.equals(authorizationRequest.getRedirectUri()) &&
+
+		if (isValidRedirectUrl(request, authorizationRequest.getRedirectUri()) &&
```

Comment:
@ttddyy This can be simplified as follows:

```java
		String requestUrl = UrlUtils.buildFullRequestUrl(request);
		MultiValueMap<String, String> params = OAuth2AuthorizationResponseUtils.toMultiMap(request.getParameterMap());
		if (requestUrl.startsWith(authorizationRequest.getRedirectUri()) &&
				OAuth2AuthorizationResponseUtils.isAuthorizationResponse(params)) {
			return true;
		}
```

Reply:
@jgrandja 
There is a problem with `startWith()` check which is caught by existing test.

Let's say here is changed to use `startWith()` as below:

```java
String requestUrl = UriComponentsBuilder.fromHttpUrl(UrlUtils.buildFullRequestUrl(request))
    .replaceQueryParam(OAuth2ParameterNames.CODE)
    .replaceQueryParam(OAuth2ParameterNames.STATE)
    .build()
    .toUriString();
MultiValueMap<String, String> params = OAuth2AuthorizationResponseUtils.toMultiMap(request.getParameterMap());

if (requestUrl.startsWith(authorizationRequest.getRedirectUri()) &&
    OAuth2AuthorizationResponseUtils.isAuthorizationResponse(params)) {
    return true;
}
return false;
```

This logic matches when:
- requestUrl = `http://localhost/callback/client-1-no-match`
- authorizationRequest.getRedirectUri() = `http://localhost/callback/client-1`

Thus, simple `startWith()` check doesn't work.

So, if we don't want to share the code between filter and validator:
- put same logic in both place
- or, maybe perform through check(`isValidRedirectUrl`) in filter, but simply only do `startWith()` check in validator

Reply:
@ttddyy Yes, you are right, there is a problem with using `startsWith()`. Instead we should perform exact matching on all parts of the URI, with the exception of the query parameters since there may be additional parameters sent in the authorization response.

In the `shouldProcessAuthorizationResponse()` method, can you make use of `UriComponentsBuilder` for both the `request` (authorization response) and `authorizationRequest.getRedirectUri()` and perform exact matching on each part, eg. scheme, userInfo, host, port, path. The query parameters (if any) in the `authorizationRequest.getRedirectUri()` should exist in the same order in `request` (authorization response).

On a side note, the validation logic for `redirect-uri` in `OAuth2AuthorizationExchangeValidator` is redundant. The matching happens in `shouldProcessAuthorizationResponse()` so we don't need to double validate in `OAuth2AuthorizationExchangeValidator`. Please remove the existing validation which will simply things.